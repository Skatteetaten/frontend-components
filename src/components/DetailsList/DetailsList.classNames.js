import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes, FontWeights } from '../utils/fonts';

function getBackgroundColor(props) {
  const { palette } = getTheme();

  if (props.background === 'white') {
    return palette.skeColor.white;
  } else {
    return 'none';
  }
}

function getBorderColor(props) {
  const { palette } = getTheme();

  if (props.background === 'white') {
    return palette.skeColor.white;
  } else {
    return palette.skeColor.lightGrey;
  }
}

function getHoverColor(props) {
  const { palette } = getTheme();

  if (props.background === 'white') {
    return palette.skeColor.neutralGrey;
  } else {
    return palette.skeColor.lightGrey;
  }
}

export const getClassNames = (props, col) => {
  const { palette } = getTheme();
  const inset = 0;
  const position = 'relative';
  const radius = '0';

  return mergeStyles({
    outline: 'transparent',
    position,
    background: palette.themePrimary,
    selectors: {
      '::-moz-focus-inner': {
        border: '0'
      },
      '.ms-Fabric.is-focusVisible.is-focusVisible & .ms-DetailsRow:focus': {
        backgroundColor: palette.themePrimary,
        color: palette.white
      },
      '.ms-Fabric.is-focusVisible.is-focusVisible & .ms-DetailsRow:focus:after': {
        content: '""',
        position: 'absolute',
        left: inset + 1,
        top: inset + 1,
        bottom: inset + 1,
        right: inset + 1,
        border: '2px solid ' + palette.themePrimary,
        borderRadius: radius,
        outline: 'transparent',
        zIndex: 1
      },
      '.ms-Fabric.is-focusVisible.is-focusVisible & .ms-DetailsHeader-cellTitle:focus:after': {
        content: '""',
        position: 'absolute',
        left: inset + 1,
        top: inset + 1,
        bottom: inset + 1,
        right: inset + 1,
        border: '2px solid ' + palette.themePrimary,
        borderRadius: radius,
        outline: 'transparent',
        zIndex: 1
      },
      '.ms-GroupHeader': {
        background: getBackgroundColor(props),
        selectors: {
          ':hover': {
            background: getHoverColor(props)
          }
        }
      },
      '.ms-GroupHeader .ms-GroupHeader-expand': {
        width: 20,
        height: 20,
        color: palette.blue,
        border: '1px solid ' + palette.blue,
        borderRadius: '50%',
        position: 'relative'
      },
      '.ms-GroupHeader .ms-GroupHeader-expand i': {
        position: 'absolute',
        transformOrigin: 'center center'
      },
      '.ms-DetailsHeader-collapseButton': {
        color: palette.skeColor.blue
      },
      '&.ms-DetailsList': {
        fontSize: FontSizes.medium,
        background: 'none'
      },
      '.ms-GroupHeader-title': {
        fontSize: FontSizes.medium,
        fontWeight: FontWeights.semibold
      },
      '& .ms-DetailsHeader-cell [data-icon-name="SortDown"], & .ms-DetailsHeader-cell [data-icon-name="SortUp"]': {
        fontSize: FontSizes.icon,
        color: palette.skeColor.blue
      },
      '& .ms-DetailsHeader-cell': {
        fontSize: FontSizes.medium,
        fontWeight: FontWeights.semibold,
        minHeight: 32,
        selectors: {
          ':hover': {
            background: getHoverColor(props)
          }
        }
      },
      '.ms-List-cell': {
        minHeight: 32 // + 2px border
      },
      //row
      '& .ms-FocusZone.ms-DetailsHeader': {
        paddingBottom: 0,
        borderBottom: '2px solid' + getBorderColor(props),
        background: getBackgroundColor(props)
      },
      '& .ms-FocusZone.ms-DetailsRow': {
        minHeight: 26,
        borderBottom: '2px solid' + getBorderColor(props),
        background: getBackgroundColor(props),
        selectors: {
          ':hover': {
            background: getHoverColor(props)
          },
          '& .is-selected': {
            background: palette.neutralQuaternaryAlt
          }
        }
      },
      //cell
      '& .ms-DetailsRow-cell': {
        fontSize: FontSizes.medium,
        padding: '8px',
        minHeight: '32px',
        selectors: {
          '& .ms-DetailsRow-cellCheck': {
            padding: 0
          }
        }
      },
      '& .ms-DetailsRow.is-selected': {
        background: palette.skeColor.neutralQuaternaryAlt
      },
      '& .ms-FocusZone .ms-DetailsRow.is-selected': {
        background: palette.skeColor.neutralQuaternaryAlt
      },
      // hack to overrride checkbox style used in Fabric detaillist
      '& .ms-DetailsRow-check': {
        height: 32,
        width: 32,
        selectors: {
          '.ms-Check': {
            width: 20,
            height: 20
          },
          '.ms-Check:before, .ms-Check.is-checked:before': {
            background: 'none',
            borderRadius: 0,
            width: 20,
            height: 20
          },
          '.ms-Check.is-checked:before': {
            background: palette.blue
          },
          '.ms-Check-circle': {
            fontFamily: 'none',
            fontSize: 0,
            height: 20,
            width: 20,
            border: '1px solid #000'
          },
          '.ms-Check-check': {
            margin: '3px 0 0 1px'
          },
          '.ms-Check-check:hover': {
            opacity: 0,
            cursor: 'pointer'
          },
          '.ms-Check.is-checked .ms-Check-check:hover': {
            opacity: 1
          }
        }
      }
    }
  });
};
