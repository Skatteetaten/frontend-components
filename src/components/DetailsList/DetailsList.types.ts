import { IDetailsListProps, IColumn as FabricIColumn } from '@fluentui/react';

export interface DetailsListProps extends IDetailsListProps {
  background?: 'white' | 'transparent';
  /**  Konfigurasjon for kolonnenavn og rekkefølge */
  columns?: Array<IColumn>;
  isSorted?: boolean;
  isSortedDescending?: boolean;
  /** Mulighet for å skru av hover-effekt dersom listeelement ikke er klikkbar. Kan ikke kombineres med checkbox */
  noHover?: boolean;
  onSortUpdate?: (...args: any[]) => any;
}

export interface IColumn extends FabricIColumn {
  sortItems?: any;
}
