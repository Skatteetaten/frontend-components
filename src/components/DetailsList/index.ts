export { default } from './DetailsList';
export * from './DetailsList';
