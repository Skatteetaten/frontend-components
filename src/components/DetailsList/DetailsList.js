import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import {
  DetailsList as FabricDetailsList,
  DetailsListLayoutMode,
  SelectionMode,
  ConstrainMode,
  CheckboxVisibility,
  Selection
} from 'office-ui-fabric-react/lib-commonjs/DetailsList';
import { getClassNames } from './DetailsList.classNames';

/**
 * @visibleName DetailsList (Sammensatt tabell)
 */
export class DetailsList extends React.PureComponent {
  static ConstrainMode = ConstrainMode;
  static SelectionMode = SelectionMode;
  static CheckboxVisibility = CheckboxVisibility;
  static DetailsListLayoutMode = DetailsListLayoutMode;
  static Selection = Selection;

  static propTypes = {
    /** Angir om detaljlisten skal være gjennomsiktig eller hvit*/

    background: PropTypes.oneOf(['white', 'transparent']),
    /** Innholdselementer i listen */
    items: PropTypes.array,
    /** Kontrollerer hvordan er kolonnene blir justert les mer [her](https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts)*/
    layoutMode: PropTypes.any,
    /** Om sjekkboks skal være synlig eller ikke */
    checkboxVisibility: PropTypes.any,
    /** Angir om man skulle velge rader vha en sjekkboks. Valg inkluderer: none, single, multiple */
    selectionMode: PropTypes.any,
    /** Avgjør hvordan liste håndterer overflow */
    constrainMode: PropTypes.any,
    /** Viser en pil hvis kolonnen er sortert */
    isSorted: PropTypes.bool,
    /** Avgjør hvilken retning sorteringspilen vises */
    isSortedDescending: PropTypes.bool,
    className: PropTypes.string,
    /** Global attributt som må være unik for hele HTML dokumentet */
    id: PropTypes.string,
    /** ({items, columns}) => void; Brukes for å oppdatere items og columns. */
    onSortUpdate: PropTypes.func
  };

  static defaultProps = {
    background: 'white',
    items: [],
    layoutMode: DetailsList.DetailsListLayoutMode.justified,
    checkboxVisibility: DetailsList.CheckboxVisibility.hidden,
    selectionMode: DetailsList.SelectionMode.none,
    constrainMode: DetailsList.ConstrainMode.unconstrained
  };

  sortColumn = sortItems => (ev, column) => {
    const { items, columns } = this.props;

    const currentColumn = columns.filter(currCol => {
      return currCol.key === column.key;
    })[0];

    const newColumns = columns.map(newCol => {
      if (newCol === currentColumn) {
        return {
          ...newCol,
          isSorted: true,
          isSortedDescending: !currentColumn.isSortedDescending
        };
      }
      return {
        ...newCol,
        isSorted: false,
        isSortedDescending: true
      };
    });

    const sortedItems = sortItems({
      isDescending: currentColumn.isSortedDescending,
      fieldName: currentColumn.fieldName,
      items: items
    });

    this.props.onSortUpdate &&
      this.props.onSortUpdate({
        columns: newColumns,
        items: sortedItems
      });
  };

  render() {
    const {
      background,
      items,
      columns,
      layoutMode,
      checkboxVisibility,
      selectionMode,
      constrainMode,
      isSorted,
      isSortedDescending,
      className,
      id,
      ...props
    } = this.props;

    const enhancedColumns = columns.map(col =>
      !col.sortItems
        ? col
        : {
            ...col,
            onColumnClick: this.sortColumn(col.sortItems)
          }
    );

    return (
      <div id={id}>
        <FabricDetailsList
          {...props}
          className={classnames(getClassNames(this.props), className)}
          items={items}
          setKey="set"
          columns={enhancedColumns}
          layoutMode={layoutMode}
          checkboxVisibility={checkboxVisibility}
          selectionMode={selectionMode}
          constrainMode={constrainMode}
          checkboxCellClassName={'DetailsListCheckbox'}
          isSorted={isSorted}
          isSortedDescending={isSortedDescending}
        />
      </div>
    );
  }
}

export default DetailsList;
