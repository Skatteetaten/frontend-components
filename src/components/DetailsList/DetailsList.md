** Detailjliste brukes til å vise store mengder informasjon som skal kunne sorteres, grupperes og filtreres **

Kolonner som kan sorteres

```js
import DetailsList from '@skatteetaten/frontend-components/DetailsList';

const initialState = {
  columns: [
    {
      key: 'column1',
      name: 'Fastsatt',
      fieldName: 'fastsatt',
      minWidth: 50,
      maxWidth: 200,
      isResizable: true,
      isSorted: true,
      isSortedDescending: false,
      sortItems: ({ isDescending, fieldName, items }) => {
        if (isDescending) {
          return items.sort((a, b) => a[fieldName].localeCompare(b[fieldName]));
        } else {
          return items.sort((a, b) => b[fieldName].localeCompare(a[fieldName]));
        }
      }
    },
    {
      key: 'column2',
      name: 'Avgiftstype',
      fieldName: 'avgiftstype',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true,
      isSortable: true,
      isSorted: false,
      isSortedDescending: false,
      headerClassName: 'is-sortable',
      sortItems: ({ isDescending, fieldName, items }) => {
        if (isDescending) {
          return items.sort((a, b) => a[fieldName].localeCompare(b[fieldName]));
        } else {
          return items.sort((a, b) => b[fieldName].localeCompare(a[fieldName]));
        }
      }
    },
    {
      key: 'column3',
      name: 'Avgiftsgruppe',
      fieldName: 'avgiftsgruppe',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column4',
      name: 'Beløp',
      fieldName: 'beloep',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    },
    {
      key: 'column5',
      name: 'Fritak',
      fieldName: 'fritak',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column6',
      name: 'Status',
      fieldName: 'status',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    }
  ],
  items: [
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6045.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.05.2018',
      avgiftstype: 'BR',
      avgiftsgruppe: '525',
      beloep: '6033.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '02.03.2018',
      avgiftstype: 'AR',
      avgiftsgruppe: '525',
      beloep: '100.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '03.03.2018',
      avgiftstype: 'DR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '04.01.2018',
      avgiftstype: 'BR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'IR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbegynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbegynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbeynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'QR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Ny',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Ny',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    }
  ]
};
<DetailsList
  columns={state.columns}
  items={state.items}
  constrainMode={DetailsList.ConstrainMode.horizontalConstrained}
  onSortUpdate={({ items, columns }) => setState({ items, columns })}
/>;
```

Listen er delt opp i rader som er gruppert

```js
import DetailsList from '@skatteetaten/frontend-components/DetailsList';

const initialState = {
  columns: [
    {
      key: 'column1',
      name: 'Fastsatt',
      fieldName: 'fastsatt',
      minWidth: 50,
      maxWidth: 200,
      isResizable: true
    },
    {
      key: 'column2',
      name: 'Avgiftstype',
      fieldName: 'avgiftstype',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column3',
      name: 'Avgiftsgruppe',
      fieldName: 'avgiftsgruppe',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column4',
      name: 'Beløp',
      fieldName: 'beloep',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    },
    {
      key: 'column5',
      name: 'Fritak',
      fieldName: 'fritak',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column6',
      name: 'Status',
      fieldName: 'status',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    }
  ],
  items: [
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbegynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbegynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Påbeynt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Ny',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Ny',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    }
  ]
};
<div className="ExampleSpacing8">
  <h3>Ekspanderbare rader</h3>
  <DetailsList
    columns={state.columns}
    items={state.items}
    groups={[
      {
        key: 'Ekspanderbar gruppe - 1',
        name: 'Gruppeoverskrift',
        startIndex: 0,
        count: 2,
        fastsatt: '31.01.2018'
      },
      {
        key: 'Ekspanderbar gruppe - 2',
        name: 'Gruppeoverskrift',
        startIndex: 2,
        count: 2
      }
    ]}
  />
</div>;
```

Brukeren kan velge en eller flere rader vha av sjekkbokser.

```js
import DetailsList from '@skatteetaten/frontend-components/DetailsList';

const initialState = {
  columns: [
    {
      key: 'column1',
      name: 'Fastsatt',
      fieldName: 'fastsatt',
      minWidth: 50,
      maxWidth: 200,
      isResizable: true
    },
    {
      key: 'column2',
      name: 'Avgiftstype',
      fieldName: 'avgiftstype',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column3',
      name: 'Avgiftsgruppe',
      fieldName: 'avgiftsgruppe',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column4',
      name: 'Beløp',
      fieldName: 'beloep',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    },
    {
      key: 'column5',
      name: 'Fritak',
      fieldName: 'fritak',
      minWidth: 50,
      maxWidth: 150,
      isResizable: true
    },
    {
      key: 'column6',
      name: 'Status',
      fieldName: 'status',
      minWidth: 100,
      maxWidth: 200,
      isResizable: true
    }
  ],
  items: [
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null,
      checkboxVisibility: true
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Under arbeid',
      kravgrunnlag: null,
      links: null
    },
    {
      fastsatt: '31.01.2018',
      avgiftstype: 'OR',
      avgiftsgruppe: '525',
      beloep: '6064.00',
      fritak: '',
      status: 'Iverksatt',
      kravgrunnlag: null,
      links: null
    }
  ]
};

function chooseRow() {
  console.log('chooseRow');
}

<div className="ExampleSpacing8">
  <h3>Valgbare rader</h3>
  <DetailsList
    marqueeSelection
    columns={state.columns}
    items={state.items}
    selectionMode={DetailsList.SelectionMode.single}
    checkboxVisibility={DetailsList.CheckBoxVisibility.always}
  />
</div>;
```

```js noeditor
import Accordion from '@skatteetaten/frontend-components/Accordion';
import AccordionItem from '@skatteetaten/frontend-components/Accordion/AccordionItem';

<Accordion>
  <AccordionItem
    isOpen
    toggleContent
    toggleButtonText={'Bruk'}
    stepId={'step-1-1'}
  >
    <p>
      DetailsList brukes når du har liste av mange repterende rader, og samtidig
      har behov for:
    </p>
    <ul>
      <li>Gruppere rader</li>
      <li>Velge/markere en eller flere rader</li>
      <li>Sortere kolonner</li>
    </ul>
    <p>
      Sørg for å ha den viktigste kolonnen først til venstre, og at det
      tilstrekkelig kolonnebredde til å vise innholdet. Dersom innholdet er tar
      mye plass - vurder å bruke variabel høyde på radene.
    </p>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Universell utforming'}
    stepId={'step-1-2'}
  >
    <ul>
      <li>Ikke bruk listen til å vise kommandoer eller innstillinger.</li>
      <li>Ikke ha for mye data slik at brukeren på scrolle horisontalt.</li>
      <li>
        Listen bør ha en vilkårlig størrelse slik at informasjonen ikke blir
        trykt sammen og uoversiktelig.
      </li>
      <li>
        Sørg for at listen har fokusmarkering slik at brukere med skjermleser
        kan tabbe seg igjennom listen.
      </li>
    </ul>
  </AccordionItem>
  <AccordionItem
    toggleContent
    toggleButtonText={'Tekniske detaljer'}
    stepId={'step-1-3'}
  >
    <p>
      <a
        href="https://developer.microsoft.com/en-us/fabric#/components/detailslist#Implementation"
        target="_blank"
      >
        Se flere tilgjengelige props i Fabric dokumentasjonen
      </a>
    </p>
  </AccordionItem>
</Accordion>;
```
