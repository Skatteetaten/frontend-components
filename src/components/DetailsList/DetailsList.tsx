import * as React from 'react';
import classnames from 'classnames';
import {
  DetailsList as FabricDetailsList,
  DetailsListLayoutMode,
  SelectionMode,
  ConstrainMode,
  CheckboxVisibility,
  Selection,
  IDetailsListProps
} from 'office-ui-fabric-react/lib-commonjs/DetailsList';
import { getClassNames } from './DetailsList.classNames';

interface DetailsListProps extends IDetailsListProps {
  background?: 'white' | 'transparent';
  isSorted?: boolean;
  isSortedDescending?: boolean;
  onSortUpdate?: (...args: any[]) => any;
}
/**
 * @visibleName DetailsList (Sammensatt tabell)
 */
export class DetailsList extends React.PureComponent<DetailsListProps, {}> {
  static ConstrainMode = ConstrainMode;
  static SelectionMode = SelectionMode;
  static CheckBoxVisibility = CheckboxVisibility;
  static DetailsListLayoutMode = DetailsListLayoutMode;
  static Selection = Selection;

  static defaultProps = {
    background: 'white',
    checkboxVisibility: DetailsList.CheckBoxVisibility.hidden,
    constrainMode: DetailsList.ConstrainMode.unconstrained,
    items: [],
    layoutMode: DetailsList.DetailsListLayoutMode.justified,
    selectionMode: DetailsList.SelectionMode.none
  };
  // @ts-ignore TODO
  sortColumn = sortItems => (ev, column) => {
    const { items, columns } = this.props;
    // @ts-ignore TODO
    const currentColumn = columns.filter(currCol => {
      return currCol.key === column.key;
    })[0];
    // @ts-ignore TODO
    const newColumns = columns.map(newCol => {
      if (newCol === currentColumn) {
        return {
          ...newCol,
          isSorted: true,
          isSortedDescending: !currentColumn.isSortedDescending
        };
      }
      return {
        ...newCol,
        isSorted: false,
        isSortedDescending: true
      };
    });

    const sortedItems = sortItems({
      isDescending: currentColumn.isSortedDescending,
      fieldName: currentColumn.fieldName,
      items: items
    });

    this.props.onSortUpdate &&
      this.props.onSortUpdate({
        columns: newColumns,
        items: sortedItems
      });
  };

  render() {
    const { background, columns, className, ...props } = this.props;
    // @ts-ignore TODO
    const enhancedColumns = columns.map(col =>
      // @ts-ignore TODO
      !col.sortItems
        ? col
        : {
            ...col,
            // @ts-ignore TODO
            onColumnClick: this.sortColumn(col.sortItems)
          }
    );

    return (
      <FabricDetailsList
        {...props}
        // @ts-ignore TODO
        className={classnames(getClassNames(this.props), className)}
        setKey="set"
        columns={enhancedColumns}
        checkboxCellClassName={'DetailsListCheckBox'}
      />
    );
  }
}

export default DetailsList;
