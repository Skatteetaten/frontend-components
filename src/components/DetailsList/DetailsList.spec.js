import React from 'react';
import toJson from 'enzyme-to-json';
import DetailsList from './DetailsList';
import { mount, shallow } from 'enzyme';

/**
 * skipViewportMeasures={true} must be set for DetailsList to render properly
 * during testing.
 */

const sortStringItems = ({ isDescending, fieldName, items }) => {
  if (isDescending) {
    return items.sort((a, b) => a[fieldName].localeCompare(b[fieldName]));
  } else {
    return items.sort((a, b) => b[fieldName].localeCompare(a[fieldName]));
  }
};

const columns = [
  {
    key: 'column1',
    name: 'Name',
    fieldName: 'name',
    isSorted: true,
    isSortedDescending: false,
    sortItems: sortStringItems
  },
  {
    key: 'column2',
    name: 'Age',
    fieldName: 'age',
    isSorted: false,
    isSortedDescending: false,
    sortItems: sortStringItems
  }
];

describe('DetailsList komponent', () => {
  it('matcher snapshot', () => {
    const items = [{ name: 'Foo', age: 21 }];
    const wrapper = shallow(<DetailsList items={items} columns={columns} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render new items when props.items changes', () => {
    const items = [{ name: 'Foo', age: 21 }];
    const wrapper = mount(
      <DetailsList
        id="test"
        items={items}
        columns={columns}
        skipViewportMeasures={true}
      />
    );
    expect(wrapper.html()).toContain('Foo');
    expect(wrapper.html()).not.toContain('Bar');

    wrapper.setProps({ items: [...items, { name: 'Bar', age: 22 }] });

    expect(wrapper.html()).toContain('Foo');
    expect(wrapper.html()).toContain('Bar');
  });

  it('should sort items when clicked on column', () => {
    const items = [{ name: 'Foo', age: 21 }, { name: 'Bar', age: 22 }];

    const wrapper = mount(
      <DetailsList
        id="test"
        items={items}
        columns={columns}
        skipViewportMeasures={true}
      />
    );

    const onSortUpdateMock = jest.fn(({ items, columns }) => {
      wrapper.setProps({ items, columns });
    });

    wrapper.setProps({
      onSortUpdate: onSortUpdateMock
    });

    const columnHeader = wrapper.find(`[data-item-key="${columns[0].key}"]`);
    const columnButton = columnHeader.find(`span[role="button"]`);

    expect(columnHeader.length).toBe(1);
    expect(columnButton.length).toBe(1);

    columnButton.simulate('click');
    expect(wrapper.prop('items')[0].name).toBe('Foo');
    expect(columnHeader.html()).toContain('aria-sort="descending"');

    columnButton.simulate('click');
    expect(wrapper.prop('items')[0].name).toBe('Bar');
    expect(columnHeader.html()).toContain('aria-sort="ascending"');

    expect(onSortUpdateMock.mock.calls.length).toBe(2);
  });
});
