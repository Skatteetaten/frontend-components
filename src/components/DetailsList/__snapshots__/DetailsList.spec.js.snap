// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsList komponent matcher snapshot 1`] = `
<StyledWithViewportComponent
  checkboxCellClassName="DetailsListCheckBox"
  checkboxVisibility={2}
  className="css-37"
  columns={
    Array [
      Object {
        "fieldName": "name",
        "isSorted": true,
        "isSortedDescending": false,
        "key": "column1",
        "name": "Name",
        "onColumnClick": [Function],
        "sortItems": [Function],
      },
      Object {
        "fieldName": "age",
        "isSorted": false,
        "isSortedDescending": false,
        "key": "column2",
        "name": "Age",
        "onColumnClick": [Function],
        "sortItems": [Function],
      },
    ]
  }
  constrainMode={0}
  items={
    Array [
      Object {
        "age": 21,
        "name": "Foo",
      },
    ]
  }
  layoutMode={1}
  selectionMode={0}
  setKey="set"
/>
`;
